proposed/resource-limit.md

```diff
@@ -0,0 +1,559 @@
+
+1. resource count <= `ResourceLimit`: `AcquireAsync` succeeds and will return a `Resource` representing the acquired resources immediately
+2. `ResourceLimit` < resource count <= `ResourceLimit`+ `MaxQueueSize`: `AcquireAsync` will enqueue or push the pending request until more resources become available.
+3. resource count > `ResourceLimit`+ `MaxQueueSize`: `AcquireAsync` will throw an `ResourceLimitExceededException` immediately if `EnqueueIncomingRequest` is configured. If `PushIncomingRequest` is configured, the oldest registration in the stack is canceled with the `ResourceLimitExceededException` exception.
+
+Side note:
+
+There are technically more queue behaviours that are possible which can be computed as the cross product of {`Queue`, `Stack`} X {`FailOldest`, `FailNewest`, `FailIncoming`}. However, it is unclear whether all combinations are useful and the two most sensible options are encoded in the `ResourceDepletedMode.EnqueueIncomingRequest` and `ResourceDepletedMode.PushIncomingRequest` enums.
+
+- `TickEventHandler`
+
+This event handler is used to allow a shared clock to trigger replenishments for multiple rate limiters instead of requiring each individual rate limiter to create a Timer to replenishing its resources.
+
+- `Window`
+
+This represents the time span where the `ResourceLimit` is to be applied.
+
+1. `SlidingWindowRateLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+- `TickEventHandler` - see `FixedWindowRateLimiterOptions`
+- `Window` - see `FixedWindowRateLimiterOptions`
+- `SegmentsPerWindow`
+
+This represents how many segments each window is to be divided into.
+
+3. `TokenBucketRateLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+- `TickEventHandler` - see `FixedWindowRateLimiterOptions`
+- `ReplenishmentPeriod`
+
+This specifies how often replenishment occurs.
+
+- `TokensPerPeriod`
+
+This specifies how many tokens are restored each time replenishment occurs.
+
+1. `ConcurrencyLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+
+#### Channels
+
+Instead of modifying the existing `BoundedChannel` and `UnboundedChannel`, the proposal here is to create a new `ResourceLimitedChannel` along with a `ResourceLimitedChannelOptions`.
+
+```c#
+namespace System.Threading.Channels
+{
+    public sealed class ResourceLimitedChannelOptions : ChannelOptions
+    {
+        public ResourceLimiter WriteLimiter { get; set; }
+    }
+
+    public sealed class ResourceLimitedChannel<T> : Channel<T> { }
```

Comment:
Nit: Do we really need `ResourceLimitedChannel<T>` in the public API surface?

Reply:
No, I don't think it needs to be exposed

Reply:
Good point, I'll update it.